contracts/contracts/system_contracts/kaiabridge/Bridge.sol

```diff
@@ -0,0 +1,400 @@
+        emit ChangeGuardian(guardian, newGuardian);
+        guardian = newGuardian;
+    }
+
+    /// @dev See {IBridge-changeJudge}
+    function changeJudge(address newJudge) public override onlyGuardian {
+        emit ChangeJudge(judge, newJudge);
+        judge = newJudge;
+    }
+
+    /// @dev See {IBridge-changeTransferTimeLock}
+    function changeTransferTimeLock(uint256 duration) public override onlyGuardian {
+        TRANSFERLOCK = duration;
+        emit ChangeTransferTimeLock(TRANSFERLOCK);
+    }
+
+    /// @dev Set mintlock duration
+    /// @param seq lcok sequence
+    /// @param time to be assigned
+    function setTransferTimeLock(uint256 seq, uint256 time) internal {
+        timelocks[seq] = block.timestamp + time;
+    }
+
+    /// @dev See {IBridge-holdClaim}
+    function holdClaim(uint256 seq) public override onlyJudge {
+        setTransferTimeLock(seq, INFINITE);
+        emit HoldClaim(seq, INFINITE);
+    }
+
+    /// @dev See {IBridge-releaseClaim}
+    function releaseClaim(uint256 seq) public override onlyGuardian {
+        setTransferTimeLock(seq, 0);
+        emit ReleaseClaim(seq, 0);
+    }
+
+    /// @dev See {IBridge-pauseBridge}
+    function pauseBridge(string calldata pauseMsg) public override onlyGuardian notPause {
+        pause = true;
+        emit BridgePause(pauseMsg);
+    }
+
+    /// @dev See {IBridge-resumeBridge}
+    function resumeBridge(string calldata resumeMsg) public override onlyGuardian inPause {
+        pause = false;
+        emit BridgeResume(resumeMsg);
+    }
+
+    /// @dev See {IBridge-isPassedTimeLockDuration}
+    function isPassedTimeLockDuration(uint256 seq) public override view returns (bool) {
+        return timelocks[seq] != 0 && timelocks[seq] < block.timestamp;
+    }
+
+    /// @dev Check if the address contains code
+    /// @param addr address to be checked
+    function isContract(address addr) internal view returns (bool) {
+        uint32 size;
+        assembly {
+            size := extcodesize(addr)
+        }
+        return (size > 0);
+    }
+
+    /// @dev See {IBridge-isProvisioned}
+    function isProvisioned(uint64 seq) public override view returns (bool) {
+        return provisions[seq].seq > 0;
+    }
+
+    /// @dev See {IBridge-isProvisionedRange}
+    function isProvisionedRange(uint64 from, uint64 to) public override view returns (uint64[] memory) {
+        // Ignore the first dummy transaction
+        if (from == 0) {
+            from = 1;
+        }
+        if (to > nProvisioned) {
+            to = uint64(nProvisioned) + 1;
+        }
+        uint64 n = uint64(to - from);
+        uint64 cnt = 0;
+        uint64[] memory temp = new uint64[](n);
+
+        for (uint64 i=from; i<to; i++) {
+            if (isProvisioned(i)) {
+                temp[cnt++] = i;
+            }
+        }
+        // fitting
+        uint64[] memory provisionedRanges = new uint64[](cnt);
```

Comment:
You can resize array directly for optimizations. (applicable to other functions)

```
uint64[] memory provisionedRanges = new uint64[](n);
uint64 cnt = 0;
for (uint64 i=from; i<to; i++) {
    if (isProvisioned(i)) {
        provisionedRanges[cnt++] = i;
     }
}

assembly {
    mstore(provisionedRanges, cnt)
}
```

Reply:
Good suggestion. However, this will increase the bytecode size. Since these are used for getter purposes, how about leaving them as they are?

Reply:
Oh really? I thought it would decrease the bytecode size. Anyway, this is optimization, so let it be.